### DISCUSSION TITLE
Problem Editing Topology - Module Seemingly Not Found in Deployment's instances.fpp File

### DISCUSSION THREAD

**Discussion by John-P-Cargille** [nasa/fprime#2127] [Q&A] [answered]
Hey All,

I'm creating a custom project simulating a sandwich shop in which an "Employee" component and a "SandwichMachine" component talk to one another, in order to work out some remaining F' confusion before I start working on actual flight software. I defined the two components mentioned above, three ports, and a number of enumerations within separate .fpp files, all within the module "SandwichModule". I successfully built the project, and created a new deployment called "GrandOpening". I was able to generate and build the deployment initially, and successfully ran "fprime-gds" before editing the deployment's topology. However, when I added `instance employee \ instance sandwichMachine` to the GrandOpening/Top/topology.fpp file and `instance employee: SandwichModule.Employee base id 0x0E00 \ queue size Default.QUEUE_SIZE \ stack size Default.STACK_SIZE \ priority 100` ... `instance sandwichMachine: SandwichModule.SandwichMachine base id 0x2100 \ queue size Default.QUEUE_SIZE` to the GrandOpening/Top/instances.fpp file, and tried to run `fprime-util build -j4` from the "GrandOpening" directory, I got the following error:

<img width="856" alt="image" src="https://github.com/nasa/fprime/assets/116765885/5e5a2a03-dd21-4286-8896-1e2077a5c97c">

As it stands, I'm simply at a loss for why the topology can't seem to detect the module's existence.
I've already done some research and troubleshooting:
- Confirmed I'm in my virtual environment
- Ascertained that the syntax "instance [instanceName]: [ModuleName].[ComponentName]" seems to be correct from looking at the FPP User's Guide's sections on modules and component instances
- Removing and remaking my deployment had no effect on the subsequently generated error message
- Wrapping the instance definition in the module (a.k.a. `module SandwichModule { instance employee etc. }`) changed the error, but only to the extent that it was upset with "Employee" instead of "SandwichModule"
- Double-checked that both components are correctly enclosed within the brackets of `module SandwichModule` within their fpp files
- Made sure it wouldn't accept `GrandOpening.Employee` (the deployment name), `SandwichShop.Employee` (the project name), or `Components.Employee` (the name of the directory the Employee component is defined in)

I can't help but wonder if this problem stems from my still-fuzzy understanding of modules; I know they are converted into C++ namespaces during implementation, but do they have any purposes beyond that? When should you define multiple modules? Are there necessary naming conventions or best practices surrounding modules that I'm missing?

I'll attach my component and deployment directories here as zip files, in case anyone wants to look through the referenced fpp files; I'd also be happy to share other parts of my project if people request them, also.

[SandwichShopComponents.zip](https://github.com/nasa/fprime/files/12021259/SandwichShopComponents.zip)
[SandwichShopDeployment.zip](https://github.com/nasa/fprime/files/12021260/SandwichShopDeployment.zip)

**Comment by thomas-bc** [comment, marked as answer]
Which version of `fprime-tools` are you running?

Without seeing your entire F´ project it's hard to tell - but from looking at your Components and Deployment, it seems you may be missing a CMake call to correctly add your components to the build system. 

Try adding the following lines to your `GrandOpening/CMakeLists.txt`, **under the `# Components and Topology` section** (replacing <path/to> of course)
```cmake
###
# Components and Topology
###
add_fprime_subdirectory("${CMAKE_CURRENT_LIST_DIR}/<path/to>/Employee/")
add_fprime_subdirectory("${CMAKE_CURRENT_LIST_DIR}/<path/to>/SandwichMachine/")
[...]
```
These `add_fprime_subdirectory` calls are what register your components in your CMake project, and allow all the autocoder work and linking everything together.

**Reply by John-P-Cargille** [reply]
Thanks! Adding the following lines:
`add_fprime_subdirectory("${CMAKE_CURRENT_LIST_DIR}/../Types/")      
add_fprime_subdirectory("${CMAKE_CURRENT_LIST_DIR}/../Ports/")
add_fprime_subdirectory("${CMAKE_CURRENT_LIST_DIR}/../Components/Employee/")
add_fprime_subdirectory("${CMAKE_CURRENT_LIST_DIR}/../Components/SandwichMachine/")`
to my GrandOpening CMakeLists.txt did in fact resolve that error. So it seems as though the deployment needs to know how to build everything that's referenced directly within it, even if that was previously built by the overall project?

Unfortunately, after adding those lines and adding my unpacketized telemetry channels to the "GrandOpeningPackets.xml" file, I am now getting a different error that I still don't know how to resolve when calling `fprime-util build -j4` within the "GrandOpening" directory:

<img width="851" alt="image" src="https://github.com/nasa/fprime/assets/116765885/50fb3895-6eb8-454a-84b2-da321e2e766d">

I tried to search within the "SandwichMachineComponentAc.cpp" file for a number of the terms given in the error message, such as ":m_p_cmdIn_in", which only appears once in the file and doesn't seem to be called as a function--as the error message would suggest--and "init(int, int)", which doesn't appear anywhere. Any ideas? Let me know if you think I should open a new thread for this issue, also.

I don't know if this is still relevant, but I'm on version 3.2.0 of `fprime-tools`

**Reply by bocchino** [reply]
The linker issue you are reporting can occur when a component definition has command ports but no command specifiers. This is a known issue with the XML-to-C++ autocoder. It will be fixed in the next version of F Prime, when we replace the XML-to-C++ autocoder with the FPP-to-C++ autocoder.

**Reply by John-P-Cargille** [reply]
Thanks so much! I would have never thought of that. Commenting out the command ports in the SandwichMachine component did in fact resolve the error, and I was able to successfully build the deployment and run the gds using it. Success! The sandwich shop is open for business!
